### PR TITLE
Feature 66 find users in the keycloak by id

### LINKED ISSUE #66: Find users in the keycloak by id

### PR DESCRIPTION
**Pull request by Iiro127**
.

### CODE REVIEW

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/PersonsController.kt [hunk 1]
```diff
@@ -125,19 +106,19 @@ class PersonsController {
...
-        }
-
         return if (active == false) {
             persons
         } else {
             filterPersons(persons)
         }
     }
 
+    suspend fun findPerson(personId: Int): ForecastPerson? {
```
Comment (Jdallos) [inline src/main/kotlin/fi/metatavu/timebank/api/controllers/PersonsController.kt]: Missing Docs

File: src/main/kotlin/fi/metatavu/timebank/api/impl/PersonsApi.kt [hunk 2]
```diff
@@ -49,6 +50,13 @@ class PersonsApi: PersonsApi, AbstractApi() {
         }
     }
 
+    override suspend fun findPerson(personId: Int): Response {
```
Comment (Jdallos) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/PersonsApi.kt]: Missing Docs

**Review by Jdallos — APPROVED** [review]
Couple of missing Docs and fine by me. Hopefully Ville can review at some point

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/PersonsController.kt [hunk 3]
```diff
@@ -125,19 +106,19 @@ class PersonsController {
...
+        val person = forecastService.findPerson(personId)
+
+        return person
+    }
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/controllers/PersonsController.kt]:
```suggestion
    suspend fun findPerson(personId: Int): ForecastPerson? {
        return forecastService.findPerson(personId)
    }
```

You can directly return with either as in the suggestion or following
`suspend fun findPerson(personId: Int): ForecastPerson? = forecastService.findPerson(personId)`

File: src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt [hunk 4]
```diff
@@ -63,6 +63,19 @@ class ForecastService {
         ).toList()
     }
 
+    /**
+     * Finds person by id from forecast
+     *
+     * @param personId Int
+     * @return ForecastPerson
+     */
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt]:
```suggestion
    /**
     * Finds person by id from Forecast
     *
     * @param personId persons Forecast id
     * @return ForecastPerson
     */
```

File: src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt [hunk 5]
```diff
@@ -63,6 +63,19 @@ class ForecastService {
...
+    fun findPerson(personId: Int): ForecastPerson? {
+        return jacksonObjectMapper().readValue(
+            doRequest("/v2/persons/$personId"),
+            ForecastPerson::class.java
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt]: Also it would probably make sense to move the deserialization process across the methods of this class to `ForecastService::doRequest`.

```kotlin
private inline fun <reified T>doRequest(path: String): T {
    return try {
        ...
        when (response.code()) {
            200 -> jacksonObjectMapper().readValue(response.body()?.string(), T::class.java)
```

Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt]: Then for example `ForecastService::findPerson` can be simplified down to

```kotlin
fun findPerson(personId: Int): ForecastPerson? {
    return doRequest("/v2/persons/$personId")
}
```

and `ForecastService::getPersons`
```kotlin
fun getPersons(): List<ForecastPerson> {
    return doRequest("/v2/persons")
}
```

and obviously same thing for rest of the methods.

File: src/main/kotlin/fi/metatavu/timebank/api/impl/PersonsApi.kt [hunk 6]
```diff
@@ -49,6 +50,13 @@ class PersonsApi: PersonsApi, AbstractApi() {
...
+        loggedUserId ?: return createUnauthorized("Invalid token!")
+        val person = personsController.findPerson(personId) ?: return createNotFound("Person not found!")
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/PersonsApi.kt]:
```suggestion
        val person = personsController.findPerson(personId) ?: return createNotFound("Person with id $personId not found!")
```

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/PersonsController.kt [hunk 7]
```diff
@@ -52,26 +52,7 @@ class PersonsController {
...
-            sunday = person.sunday,
-            active = person.active,
-            unspentVacations = person.unspentVacations,
-            spentVacations = person.spentVacations,
-            minimumBillableRate = keycloakController.getUsersMinimumBillableRate(keycloakUser),
-            language = person.language,
-            startDate = person.startDate,
-            keycloakId = person.keycloakId
-        )
+        return person.copy(minimumBillableRate = keycloakController.getUsersMinimumBillableRate(keycloakUser))
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/controllers/PersonsController.kt]: Now that we have means of finding just one person from Forecast, it would make sense to, instead of returning the same `Person` object we received, to find the person from Forecast again and return that. IMO that more properly mimics a scenario where Persons would be stored in our own database and ensure that the front-end always has up-to-date persons.

File: src/main/kotlin/fi/metatavu/timebank/api/impl/translate/PersonsTranslator.kt [hunk 8]
```diff
@@ -16,11 +17,15 @@ class PersonsTranslator: AbstractTranslator<ForecastPerson, Person>() {
     @Inject
     lateinit var keycloakController: KeycloakController
 
+    @Inject
+    lateinit var vacationUtils: VacationUtils
 
-    override fun translate(entity: ForecastPerson): Person {
+    override suspend fun translate(entity: ForecastPerson): Person {
         val keycloakUser = keycloakController.findUserByEmail(entity.email)
         val minimumBillableRate = if (keycloakUser == null) 75 else keycloakController.getUsersMinimumBillableRate(keycloakUser)
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/translate/PersonsTranslator.kt]: Move the default minimum billable rate (75) to a static constant within `PersonsController`.

In Kotlin, static class members/methods are defined as
```kotlin
class PersonsController {
    ...
    companion object {
        const val DEFAULT_MINIMUM_BILLABLE_RATE = 75
    }
}
```

That would then be accessed, for example in here as `PersonsController.DEFAULT_MINIMUM_BILLABLE_RATE`.

**Review by villejuutila — CHANGES_REQUESTED** [review]
Also as an addition to these changes, Google Cloud Build is failing.

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/PersonsController.kt [hunk 9]
```diff
@@ -36,6 +38,9 @@ class PersonsController {
     @Inject
     lateinit var keycloakController: KeycloakController
 
+    @Inject
+    lateinit var personsTranslator: PersonsTranslator
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/controllers/PersonsController.kt]: I'd leave this whole class as is (e.g. revert all the changes made in this PR, as this is deprecated after this PR.

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+import javax.inject.Inject
+
+@ApplicationScoped
+class UsersController {
+
+    @Inject
+    lateinit var keycloakController: KeycloakController
+
+    @Inject
+    lateinit var userTranslator: UserTranslator
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt]: Performing translating into REST-entities in the implementation layer, `UsersApi` in this case.

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+
+    fun findUser(userId: UUID): User? {
+        val user = keycloakController.findUserById(userId)
+        if (user != null) {
+            return userTranslator.translate(user)
+        }
+
+        return null
+    }
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt]:
```suggestion
    fun findUser(userId: UUID): UserRepresentation? {
        return keycloakController.findUserById(userId)
    }
```

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt [hunk 12]
```diff
@@ -0,0 +1,36 @@
...
+
+    fun listUsers(): List<User>? {
+        val users = keycloakController.getUsersResource()
+        if (users != null) {
+            return userTranslator.translate(keycloakController.getUsersResource()!!)
+        }
+
+        return null
+    }
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt]: Return a list of `UserRepresentation`s from `KeycloakController::GetUsersResource` and then pass that through this method e.g.

```kotlin
    fun listUsers(): List<User> {
        return keycloakController.listUsers() 
    }
```

At the moment it seems like this class doesn't serve that big of a purpose but since we know it'll start to handle integrations in the next coming PR's I'd probably write this as of now as well. :)

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/VacationRequestController.kt [hunk 13]
```diff
@@ -37,6 +37,7 @@ class VacationRequestController {
                 message = vacationRequest.message,
                 createdAt = vacationRequest.createdAt,
                 updatedAt = vacationRequest.updatedAt,
+                draft = vacationRequest.draft
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/controllers/VacationRequestController.kt]: According to the issue this PR is resolving, nothing related to `VacationRequests` should've changed in this PR?

It seems to me that this branch has originally been based on the earlier PR but that base branch had some more work done before merging.
Therefore I recommend running `git fetch` > `git merge origin/develop` and resolving the conflicts in your IDE. As you can see in this PR, Github is warning you that it has conflicts with the `develop` branch.

File: src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt [hunk 14]
```diff
@@ -43,14 +44,27 @@ class ForecastService {
             val response = client.newCall(request).execute()
             when (response.code()) {
                 200 -> response.body()?.string()
-                else -> throw Error("Couldn't reach Forecast API.")
+                else -> throw Error("Couldn't reach Severa API.")
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt]: This class is called `ForecastService`. Revert this change.

File: src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt [hunk 15]
```diff
@@ -43,14 +44,27 @@ class ForecastService {
...
         }
     }
 
+    /**
+     * Finds ForecastPerson from Forecast by id
+     *
+     * @param personId Int
+     * @return ForecastPerson?
+     */
+    fun findPerson(personId: Int): ForecastPerson {
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt]: Since this whole `Persons` resource is being deprecated in this PR and given the fact that our integration with this service ends in 7 days (at the time of writing this) I'd say we don't want to do **any** work in this class.

File: src/main/kotlin/fi/metatavu/timebank/api/impl/PersonsApi.kt [hunk 16]
```diff
@@ -21,6 +22,16 @@ class PersonsApi: PersonsApi, AbstractApi() {
...
+    override suspend fun findPerson(personId: Int): Response {
+        loggedUserId ?: return createUnauthorized("Invalid token!")
+        val person = personsController.findPerson(personId)
+
+        return if (person != null) {
+            createOk(entity = person)
+        } else {
+            createNotFound("Person with id $personId not found!")
+        }
+    }
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/PersonsApi.kt]: Given that we've deprecated the `Persons` resource in this PR, I'd say don't implement this new method. If it is included in the spec I'd love to see PR for the specs where this endpoint is removed.

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt [hunk 17]
```diff
@@ -0,0 +1,36 @@
+package fi.metatavu.timebank.api.controllers
+
+import fi.metatavu.timebank.api.impl.translate.UserTranslator
+import fi.metatavu.timebank.api.keycloak.KeycloakController
+import fi.metatavu.timebank.model.User
+import java.util.UUID
+import javax.enterprise.context.ApplicationScoped
+import javax.inject.Inject
+
+@ApplicationScoped
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt:12]: This project does use class and method level documentation. Please include them, even though both the class and its methods are rather self documenting.

File: src/main/kotlin/fi/metatavu/timebank/api/impl/UsersApi.kt [hunk 18]
```diff
@@ -0,0 +1,37 @@
...
+
+import fi.metatavu.timebank.api.controllers.UsersController
+import fi.metatavu.timebank.spec.UsersApi
+import java.util.*
+import javax.enterprise.context.RequestScoped
+import javax.inject.Inject
+import javax.ws.rs.core.Response
+
+@RequestScoped
+class UsersApi: UsersApi, AbstractApi() {
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/UsersApi.kt:15]: It has been a convention in this project that API implementation has class level documentation, please include it. Even if `Implementation for Users API` does seem self documenting.

File: src/main/kotlin/fi/metatavu/timebank/api/impl/UsersApi.kt [hunk 19]
```diff
@@ -0,0 +1,37 @@
...
+    override suspend fun findUser(userId: UUID): Response {
+        loggedUserId ?: return createUnauthorized("Invalid token!")
+        val user = usersController.findUser(userId)
+
+        return if (user != null) {
+            createOk(entity = user)
+        } else {
+            createNotFound("Person with id $userId not found!")
+        }
+    }
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/UsersApi.kt]:
```suggestion
    override suspend fun findUser(userId: UUID): Response {
        loggedUserId ?: return createUnauthorized("Invalid token!")
        val user = usersController.findUser(userId) ?: return createNotFound("Person with id $userId not found!")

        createOk(entity = user)
    }
```

File: src/main/kotlin/fi/metatavu/timebank/api/impl/UsersApi.kt [hunk 20]
```diff
@@ -0,0 +1,37 @@
...
+
+    override suspend fun listUsers(): Response {
+        loggedUserId ?: return createUnauthorized("Invalid token!")
+
+        return try {
+            val persons = usersController.listUsers() ?: return createNotFound("No persons found!")
+            createOk(entity = persons)
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/UsersApi.kt]: See comments in `UsersController`. `listUsers` in both layers should return a list, no matter if the list is empty or not.

File: src/main/kotlin/fi/metatavu/timebank/api/impl/VacationRequestsApi.kt [hunk 21]
```diff
@@ -33,13 +33,14 @@ class VacationRequestsApi: VacationRequestsApi, AbstractApi() {
...
 
-    override suspend fun listVacationRequests(personId: UUID?, before: LocalDate?, after: LocalDate?): Response {
+    override suspend fun listVacationRequests(personId: UUID?, before: LocalDate?, after: LocalDate?, draft: Boolean?): Response {
         loggedUserId ?: return createUnauthorized("Invalid token!")
 
         val vacationRequests = vacationRequestController.listVacationRequests(
             personId = personId,
             before = before,
-            after = after
+            after = after,
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/VacationRequestsApi.kt]: See comments in `VacationRequestsController`.

File: src/main/kotlin/fi/metatavu/timebank/api/impl/translate/UserTranslator.kt [hunk 22]
```diff
@@ -0,0 +1,44 @@
...
+import javax.enterprise.context.ApplicationScoped
+import javax.inject.Inject
+
+@ApplicationScoped
+class UserTranslator: AbstractTranslator<UserRepresentation, User>() {
+    @Inject
+    lateinit var keycloakController: KeycloakController
+
+    override fun translate(entity: UserRepresentation): User {
+        val keycloakUser = keycloakController.findUserByEmail(entity.email)
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/translate/UserTranslator.kt]: This seems like a weird setup. Why would you want to re-find te user in this case?

Why not just 
```kotlin
    override fun translate(entity: UserRepresentation): User {
        return User(
            id = UUID.fromString(entity.id),
            ...
    }
```

File: src/main/kotlin/fi/metatavu/timebank/api/impl/translate/UserTranslator.kt [hunk 23]
```diff
@@ -0,0 +1,44 @@
...
+        return User(
+            id = keycloakUser?.let { UUID.fromString(keycloakUser.id) }!!,
+            email = entity.email,
+            firstName = entity.firstName,
+            lastName = entity.lastName,
+            isActive = true
+        )
+    }
+
+    fun translate(entities: UsersResource): List<User> {
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/translate/UserTranslator.kt]: IMO this just adds un-necessary complexity. You could just convert the `UsersResource` to a list of `UserRepresentation` before this class and use the other `::translate`.

File: src/main/kotlin/fi/metatavu/timebank/api/impl/translate/VacationRequestTranslator.kt [hunk 24]
```diff
@@ -20,6 +20,7 @@ class VacationRequestTranslator: AbstractTranslator<VacationRequest, fi.metatavu
             type = entity.type!!,
             createdAt = entity.createdAt!!,
             updatedAt = entity.updatedAt!!,
+            draft = entity.draft!!
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/translate/VacationRequestTranslator.kt:23]: See earlier comments on this matter.

File: src/main/kotlin/fi/metatavu/timebank/api/persistence/model/VacationRequest.kt [hunk 25]
```diff
@@ -40,4 +40,7 @@ class VacationRequest(
 
     @Column
     var updatedAt: OffsetDateTime? = null,
+
+    @Column
+    var draft: Boolean? = null
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/persistence/model/VacationRequest.kt]: See earlier comments on this matter.

File: src/main/kotlin/fi/metatavu/timebank/api/persistence/repositories/VacationsRequestsRepository.kt [hunk 26]
```diff
@@ -20,7 +20,7 @@ class VacationsRequestsRepository: AbstractRepository<VacationRequest, UUID>() {
      * @param after LocalDate to retrieve requests after given date
      * @return List of VacationRequests
      */
-    suspend fun listVacationRequest(personId: UUID?, before: LocalDate?, after: LocalDate?): List<VacationRequest> {
+    suspend fun listVacationRequest(personId: UUID?, before: LocalDate?, after: LocalDate?, draft: Boolean?): List<VacationRequest> {
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/persistence/repositories/VacationsRequestsRepository.kt]: See earlier comments on this matter.

File: src/main/resources/db/changeLog.xml [hunk 27]
```diff
@@ -143,4 +143,12 @@
             <column name="updatedBy" type="binary(16)"/>
         </createTable>
     </changeSet>
+
+    <changeSet id="addDraftColumnToVacationRequest" author="Veikko Vilpponen">
+        <addColumn tableName="VacationRequest">
```
Comment (villejuutila) [inline src/main/resources/db/changeLog.xml]: See earlier comments on this matter.

Hold on, how on earth is @VeikkoVilpponen authoring this? Considering he hasn't worked on the project for I don't know how long ago since. :)

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/impl/VacationRequestTestBuilderResource.kt [hunk 28]
```diff
@@ -46,11 +46,12 @@ class VacationRequestTestBuilderResource(
      * @param after optional after date
      * @return List of VacationRequests
      */
-    fun listVacationRequests(personId: UUID? = null, before: String? = null, after: String? = null): Array<VacationRequest> {
+    fun listVacationRequests(personId: UUID? = null, before: String? = null, after: String? = null, draft: Boolean? = null): Array<VacationRequest> {
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/impl/VacationRequestTestBuilderResource.kt]: See earlier comments on this matter.

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/VacationRequestsTest.kt [hunk 29]
```diff
@@ -141,6 +142,7 @@ class VacationRequestsTest: AbstractTest() {
                 message = "Lomaa!!!",
                 createdAt = getODT(getThirtyDaysAgoThirdWeek()[1].atStartOfDay()),
                 updatedAt = getODT(getThirtyDaysAgoThirdWeek()[1].atStartOfDay()),
+                draft = false
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/VacationRequestsTest.kt:150]: See earlier comments on this matter.

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/VacationRequestsTest.kt [hunk 30]
```diff
@@ -44,6 +44,7 @@ class VacationRequestsTest: AbstractTest() {
     message = "Lomaa!!!",
     createdAt = getODT(getThirtyDaysAgoThirdWeek()[1].atStartOfDay()),
     updatedAt = getODT(getThirtyDaysAgoThirdWeek()[1].atStartOfDay()),
+    draft = false
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/VacationRequestsTest.kt:47]: See earlier comments on this matter.

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/VacationRequestStatusTest.kt [hunk 31]
```diff
@@ -46,6 +46,7 @@ class VacationRequestStatusTest: AbstractTest() {
     message = "Vacation!!!",
     createdAt = getODT(getThirtyDaysAgoThirdWeek()[1].atStartOfDay()),
     updatedAt = getODT(getThirtyDaysAgoThirdWeek()[1].atStartOfDay()),
+    draft = false
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/VacationRequestStatusTest.kt:49]: See earlier comments on this matter.

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/UsersTest.kt [hunk 32]
```diff
@@ -0,0 +1,28 @@
...
+@QuarkusTestResource.List(
+    QuarkusTestResource(TestWiremockResource::class)
+)
+@TestProfile(LocalTestProfile::class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class UsersTest: AbstractTest() {
+
+    @Test
+    fun listActiveUsers() {
+        assertTrue(true)
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/UsersTest.kt]: Add tests.

**Review by villejuutila — CHANGES_REQUESTED** [review]

**Comment by Iiro127** [comment]
Resolves #77 
Resolves #72

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/PersonsController.kt (hunk 9 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/timebank/api/controllers/PersonsController.kt]: This class does seem to still have changes?

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt (hunk 10 above)
Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt]: This isn't used anywhere, so remove it. Also remove the unused imports. Your IDE should warn you about unused variables and imports before committing? 🤔

File: src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt [hunk 33]
```diff
@@ -0,0 +1,42 @@
...
+     */
+    fun findUser(userId: UUID): UserRepresentation? {
+        return keycloakController.findUserById(userId)
+    }
+
+    /**
+     * Lists Users found in Keycloak
+     *
+     * @return List<User>
+     */
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt:32]:
```suggestion
    /**
     * Lists Users found in Keycloak
     *
     * @return List of UserRepresentations
     */
```

Comment (villejuutila) [reply src/main/kotlin/fi/metatavu/timebank/api/controllers/UsersController.kt:32]: JavaDoc doesn't need types, the JavaDoc tool itself parses the types from code. IMO my suggestion is more human readable.

File: src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt [hunk 34]
```diff
@@ -1,6 +1,7 @@
 package fi.metatavu.timebank.api.forecast
 
 import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
+import com.fasterxml.jackson.module.kotlin.readValue
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/forecast/ForecastService.kt]: Isn't this import unused?

File: src/main/kotlin/fi/metatavu/timebank/api/impl/PersonsApi.kt [hunk 35]
```diff
@@ -1,6 +1,7 @@
 package fi.metatavu.timebank.api.impl
 
 import fi.metatavu.timebank.api.controllers.PersonsController
+import fi.metatavu.timebank.api.forecast.models.ForecastPerson
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/PersonsApi.kt]: Isn't this import unused?

File: src/main/kotlin/fi/metatavu/timebank/api/impl/UsersApi.kt [hunk 36]
```diff
@@ -0,0 +1,41 @@
...
+
+    @Inject
+    lateinit var usersController: UsersController
+
+    @Inject
+    lateinit var userTranslator: UserTranslator
+
+    override suspend fun findUser(userId: UUID): Response {
+        loggedUserId ?: return createUnauthorized("Invalid token!")
+        val user = usersController.findUser(userId) ?: return createNotFound("Person with id $userId not found!")
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/impl/UsersApi.kt]:
```suggestion
        val user = usersController.findUser(userId) ?: return createNotFound("User with id $userId not found!")
```

File: src/main/kotlin/fi/metatavu/timebank/api/keycloak/KeycloakController.kt [hunk 37]
```diff
@@ -48,6 +50,44 @@ class KeycloakController {
...
+     *
+     * @param user UserRepresentation
+     * @return Int unspentVacations
+     */
+    fun getUsersUnspentVacationDays(user: UserRepresentation): Int {
+        return try {
+            user.attributes["unspentVacations"]!!.first()!!.toInt()
+        } catch (e: Exception){
+            updateUsersUnspentVacationDays(user, -1)
+            -1
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/keycloak/KeycloakController.kt]: I believe @Jdallos initially requested these default values to be the ones calculated by the utilities provided by `VacationUtils`?

File: src/main/kotlin/fi/metatavu/timebank/api/keycloak/KeycloakController.kt [hunk 38]
```diff
@@ -97,6 +175,25 @@ class KeycloakController {
         return foundRealm.users()
     }
 
+    /**
+     * Searches Keycloak users
+     *
+     * @return List<UserRepresentation>
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/keycloak/KeycloakController.kt]: See other comment regarding JavaDocs

File: src/main/kotlin/fi/metatavu/timebank/api/keycloak/KeycloakController.kt [hunk 39]
```diff
@@ -67,6 +107,44 @@ class KeycloakController {
...
+            usersResource?.update(user)
+        }
+    }
+
+    /**
+     * Updates Persons spentVacations attribute
+     *
+     * @param user UserRepresentation
+     * @param newSpent Int
+     * @return Int spentVacations
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/keycloak/KeycloakController.kt]: See other comment regarding JavaDocs

File: src/main/kotlin/fi/metatavu/timebank/api/keycloak/KeycloakController.kt [hunk 40]
```diff
@@ -48,6 +50,44 @@ class KeycloakController {
...
+            -1
+        }
+    }
+
+    /**
+     * Gets spentVacations attribute for Person
+     * If not set will set and return default value of -1
+     *
+     * @param user UserRepresentation
+     * @return Int spentVacations
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/keycloak/KeycloakController.kt]: See other comment regarding JavaDocs

File: src/main/kotlin/fi/metatavu/timebank/api/keycloak/KeycloakController.kt [hunk 41]
```diff
@@ -48,6 +50,44 @@ class KeycloakController {
...
+    fun getUsersSpentVacationDays(user: UserRepresentation): Int {
+        return try {
+            user.attributes["spentVacations"]!!.first()!!.toInt()
+        } catch (e: Exception){
+            updateUsersSpentVacationDays(user, -1)
+            -1
+        }
+    }
+
+    fun updateAll(user: UserRepresentation, newMinimumBillable: Int, newUnspent: Int, newSpent: Int) {
```
Comment (villejuutila) [inline src/main/kotlin/fi/metatavu/timebank/api/keycloak/KeycloakController.kt]: Missing docs.

Also I don't see where this method is being used? If it's not used, remove it.

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/impl/UsersTestBuilderResource.kt [hunk 42]
```diff
@@ -0,0 +1,54 @@
...
+        ApiClient.accessToken = accessTokenProvider?.accessToken
+        return UsersApi(ApiTestSettings.apiBasePath)
+    }
+
+    fun getUsers(): Array<User>{
+        println("Fetching users...")
+        val response = api.listUsers()
+        println("Response: $response")
+        return response
+    }
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/impl/UsersTestBuilderResource.kt]:
```suggestion
    fun getUsers(): Array<User>{
        return api.listUsers()
    }
```

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/impl/UsersTestBuilderResource.kt [hunk 43]
```diff
@@ -0,0 +1,54 @@
...
+    fun findUser(userId: UUID): User {
+        return api.findUser(userId)
+    }
+
+    /**
+     * Asserts that listing users with a null token fails with given status
+     *
+     * @param expectedStatus expected status
+     */
+    fun assertListFailWithNullToken(expectedStatus: Int) {
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/impl/UsersTestBuilderResource.kt]: This method itself doesn't imply what kind of token, if any, it is using. Rename to `assertListFailWithStatus`

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/impl/UsersTestBuilderResource.kt (hunk 42 above)
Comment (villejuutila) [reply src/test/kotlin/fi/metatavu/timebank/api/test/functional/impl/UsersTestBuilderResource.kt]: Also this method seems to be missing docs

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/impl/UsersTestBuilderResource.kt [hunk 44]
```diff
@@ -0,0 +1,54 @@
...
+    }
+
+    fun getUsers(): Array<User>{
+        println("Fetching users...")
+        val response = api.listUsers()
+        println("Response: $response")
+        return response
+    }
+
+    fun findUser(userId: UUID): User {
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/impl/UsersTestBuilderResource.kt:43]: Also this method seems to be missing docs

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/PersonsTest.kt [hunk 45]
```diff
@@ -142,6 +143,65 @@ class PersonsTest: AbstractTest() {
         }
     }
 
+    /**
+     * Tests overriding vacation days for users via v/1/persons/{personId} -endpoint PUT method. Only admin-roles should be allowed to perform this action.
+     */
+    @Test
+    fun overrideVacationDaysTest() {
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/PersonsTest.kt]: Isn't this new test out of the scope of this PR?

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/UsersTest.kt [hunk 46]
```diff
@@ -0,0 +1,47 @@
...
+    @BeforeEach
+    fun resetScenariosBeforeEach(){
+        resetScenarios()
+    }
+
+    @Test
+    fun listUsers(){
+        createTestBuilder().use { testBuilder ->
+            val users = testBuilder.manager.users.getUsers()
+            println("Users received: ${users.contentToString()}")
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/UsersTest.kt]:
```suggestion
```

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/UsersTest.kt [hunk 47]
```diff
@@ -0,0 +1,47 @@
...
+@TestProfile(LocalTestProfile::class)
+@TestInstance(TestInstance.Lifecycle.PER_CLASS)
+class UsersTest: AbstractTest() {
+
+    /**
+     * Resets Wiremock scenario states before each test
+     */
+    @BeforeEach
+    fun resetScenariosBeforeEach(){
+        resetScenarios()
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/UsersTest.kt]: Wiremock isn't being used by this test class (at least for the time being). This is therefore redundant.

File: src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/UsersTest.kt [hunk 48]
```diff
@@ -0,0 +1,47 @@
...
+        createTestBuilder().use { testBuilder ->
+            val users = testBuilder.manager.users.getUsers()
+            println("Users received: ${users.contentToString()}")
+            val testerAUser = users.find { it.firstName == "Manager" }
+
+            assertEquals(6, users.size)
+            assertEquals(testerAUser!!.email, "manager@timebank.com")
+        }
+    }
+
```
Comment (villejuutila) [inline src/test/kotlin/fi/metatavu/timebank/api/test/functional/tests/UsersTest.kt:31]: Test for `/v1/users/{userId}` -endpoint seems to be missing.

**Review by villejuutila — CHANGES_REQUESTED** [review]

**Review by villejuutila — APPROVED** [review]
